libpod/runtime_img.go

```diff
@@ -854,50 +834,52 @@ func (r *Runtime) UntagImage(image *storage.Image, tag string) (string, error) {
 
 // RemoveImage deletes an image from local storage
 // Images being used by running containers can only be removed if force=true
-func (r *Runtime) RemoveImage(image *storage.Image, force bool) (string, error) {
+func (r *Runtime) RemoveImage(image *image.Image, force bool) error {
 	r.lock.Lock()
 	defer r.lock.Unlock()
 
 	if !r.valid {
-		return "", ErrRuntimeStopped
+		return ErrRuntimeStopped
 	}
 
 	// Get all containers, filter to only those using the image, and remove those containers
 	ctrs, err := r.state.AllContainers()
 	if err != nil {
-		return "", err
+		return err
 	}
 	imageCtrs := []*Container{}
 	for _, ctr := range ctrs {
-		if ctr.config.RootfsImageID == image.ID {
+		if ctr.config.RootfsImageID == image.ID() {
 			imageCtrs = append(imageCtrs, ctr)
 		}
 	}
-	if len(imageCtrs) > 0 && len(image.Names) <= 1 {
+	if len(imageCtrs) > 0 && len(image.Names()) <= 1 {
 		if force {
 			for _, ctr := range imageCtrs {
 				if err := r.removeContainer(ctr, true); err != nil {
-					return "", errors.Wrapf(err, "error removing image %s: container %s using image could not be removed", image.ID, ctr.ID())
+					return errors.Wrapf(err, "error removing image %s: container %s using image could not be removed", image.ID, ctr.ID())
 				}
 			}
 		} else {
-			return "", fmt.Errorf("could not remove image %s as it is being used by %d containers", image.ID, len(imageCtrs))
+			return fmt.Errorf("could not remove image %s as it is being used by %d containers", image.ID, len(imageCtrs))
 		}
 	}
 
-	if len(image.Names) > 1 && !force {
-		return "", fmt.Errorf("unable to delete %s (must force) - image is referred to in multiple tags", image.ID)
+	if len(image.Names()) > 1 && !force {
+		return fmt.Errorf("unable to delete %s (must force) - image is referred to in multiple tags", image.ID)
 	}
 
-	// If it is forced, we have to untag the image so that it can be deleted
-	if err = r.store.SetNames(image.ID, image.Names[:0]); err != nil {
-		return "", err
+	if len(image.Names()) > 1 && force {
+		// If it is forced, we have to untag the image so that it can be deleted
+		if err = r.store.SetNames(image.ID(), image.Names()[:0]); err != nil {
+			return err
+		}
 	}
-	_, err = r.store.DeleteImage(image.ID, true)
-	if err != nil {
-		return "", err
+
+	if err = image.Remove(force); err != nil {
```

Comment:
If we have an image.Remove(), shouldn't all of this happen in there instead?

Reply:
No because we dont know about containers in the image library.  so we deal with containers here and then do the removal from the image library.

Reply:
Good point, sorry for the runaround on this one.